packages/flutter/lib/src/widgets/animated_switcher.dart

```diff
@@ -292,39 +310,24 @@ class _AnimatedSwitcherState extends State<AnimatedSwitcher> with TickerProvider
     if (_currentChild != null) {
       _currentChild.controller.dispose();
     }
-    for (_AnimatedSwitcherChildEntry child in _children) {
+    for (_AnimatedSwitcherChildEntry child in _previousChildren) {
       child.controller.dispose();
     }
     super.dispose();
   }
 
-  bool get hasNewChild => widget.child != null;
-  bool get hasOldChild => _currentChild != null;
-
   @override
   void didUpdateWidget(AnimatedSwitcher oldWidget) {
     super.didUpdateWidget(oldWidget);
-    if (hasNewChild != hasOldChild || hasNewChild &&
-        !Widget.canUpdate(widget.child, _currentChild.widgetChild)) {
+    final bool hasNewChild = widget.child != null;
+    final bool hasOldChild = _currentChild != null;
+    if (hasNewChild != hasOldChild || hasNewChild && !Widget.canUpdate(widget.child, _currentChild.widgetChild)) {
       _addEntry(animate: true);
-    } else {
-      if (_currentChild != null) {
-        _currentChild.widgetChild = widget.child;
```

Comment:
if you remove this line, then surely you'll be making it impossible to update the child (in the case where it's not changing identity).

e.g. here's a test to add: create an AnimatedSwitcher, and set the child to `Text('1')`. Then, pump an isomorphic tree of widgets, but with the child set to `Text('2')`. Verify that the first time, the tree rendered `1`, and the second time it rendered `2` with no animation (and no `1`).